Dependencies/Utility/Utility/interlocked_compat.h

```diff
@@ -0,0 +1,143 @@
+
+#elif defined(__FreeBSD__) || defined(__DragonFly__)
+#include <machine/atomic.h>
+// Appears to have no native atomic operations for 16, 64 bit integers.
+inline AtomicType32 Interlocked_Increment_32(volatile AtomicType32 *addend) { return atomic_fetchadd_32(addend, AtomicType32{ 1 }); }
+
+inline AtomicType32 Interlocked_Decrement_32(volatile AtomicType32 *addend) { return atomic_fetchadd_32(addend, ~AtomicType32{ 0 }); }
+
+#elif defined(__NetBSD__)
+#include <sys/atomic.h>
+// Appears to have no native atomic operations for 16 bit integers.
+inline AtomicType32 Interlocked_Increment_32(volatile AtomicType32 *addend) { return atomic_inc_32_nv(addend); }
+inline AtomicType64 Interlocked_Increment_64(volatile AtomicType64 *addend) { return atomic_inc_64_nv(addend); }
+
+inline AtomicType32 Interlocked_Decrement_32(volatile AtomicType32 *addend) { return atomic_dec_32_nv(addend); }
+inline AtomicType32 Interlocked_Decrement_64(volatile AtomicType64 *addend) { return atomic_dec_64_nv(addend); }
+
+#elif defined(_WIN32)
+#include <windows.h>
+#if !(defined(_MSC_VER) && _MSC_VER < 1300)
+inline AtomicType16 Interlocked_Increment_16(volatile AtomicType16 *addend) { return InterlockedIncrement16(addend); }
+inline AtomicType32 Interlocked_Increment_32(volatile AtomicType32 *addend) { return InterlockedIncrement(addend); }
+inline AtomicType64 Interlocked_Increment_64(volatile AtomicType64 *addend) { return InterlockedIncrement64(addend); }
+
+inline AtomicType16 Interlocked_Decrement_16(volatile AtomicType16 *addend) { return InterlockedDecrement16(addend); }
+inline AtomicType32 Interlocked_Decrement_32(volatile AtomicType32 *addend) { return InterlockedDecrement(addend); }
+inline AtomicType64 Interlocked_Decrement_64(volatile AtomicType64 *addend) { return InterlockedDecrement64(addend); }
+#endif // !(_MSC_VER < 1300)
+
+#else
+#error platform not supported
+#endif
+
+
+// Interlocked helper functions
+
+// VC6 compatible overloaded interlocked functions - No native 16 or 64 bit functions supported, use with care
+#if defined(_MSC_VER) && _MSC_VER < 1300
+
+inline AtomicType16 interlockedIncrement(volatile AtomicType16 *addend) { return InterlockedIncrement((AtomicType32*)addend); }
+inline AtomicType32 interlockedIncrement(volatile AtomicType32 *addend) { return InterlockedIncrement(addend); }
+inline AtomicType64 interlockedIncrement(volatile AtomicType64 *addend) { return InterlockedIncrement((AtomicType32*)addend); }
+
+inline AtomicType16 interlockedDecrement(volatile AtomicType16 *addend) { return InterlockedDecrement((AtomicType32*)addend); }
+inline AtomicType32 interlockedDecrement(volatile AtomicType32 *addend) { return InterlockedDecrement(addend); }
+inline AtomicType64 interlockedDecrement(volatile AtomicType64 *addend) { return InterlockedDecrement((AtomicType32*)addend); }
+
+#else
+
+template<typename Type> inline Type interlockedIncrement(volatile Type *addend);
```

Comment:
Can add a comment here and explain that these are written as templates, so that they do not cause compile errors when one of Interlocked_Increment_16, Interlocked_Increment_32, Interlocked_Increment_64 is missing, as long as the template is not called.